# Individual Assignment A1

This assessment is marked out of 50 and comprises 50% of the final course mark.

Due by 23:59 on Monday November 18, to be submitted via email to k.zygalakis@ed.ac.uk.

### Academic misconduct

The assessment is primarily summative in nature. You are expected to be aware of and abide by University policies on academic misconduct.

- [School of Mathematics academic misconduct advice and policies](https://teaching.maths.ed.ac.uk/main/undergraduate/studies/assessment/academic-misconduct)
- [Academic Services academic misconduct information](https://www.ed.ac.uk/academic-services/students/conduct/academic-misconduct)

**This is an individual assignment - do not copy the work of another student.**

If you use any resources (e.g. textbooks or websites) then include appropriate references in your solutions. Course materials do not need to be referenced, but you should clearly state which results you are using.


### Code commentary

Your code should be extensively commented, with the functionality of each line of code explained with a comment. This is to test your understanding of the code you have written. Up to half of the marks associated with the coding part of a question may be deducted for a missing, incomplete, or inaccurate code commentary.

Your comments should explain what the code does, as well as why it does it.

The following provides an example of the expected level of commenting.

In [ ]:
def is_prime(n):
    """
    Return whether an input positive integer is prime.
    """
    
    if n == 1:        # If n is 1 ...
        return False  # ... then n is not prime
    
    for i in range(2, n):  # Test integers i from 2 to n - 1 inclusive
        if n % i == 0:     # If n is divisible by i ...
            return False   # ... then n is not prime
    # If n is not divisible by any integers from 2 to n - 1 inclusive then n is
    # prime
    return True

### Output 

Your code must generate and display all relevant output when run. Rerun your code cells after editing your code, to make sure that the output is updated.

### Markdown cells

You can enter your answers to theoretical questions in the Markdown cells provided in this notebook. To start editing the cell, press shift+enter or double click on it. You can use basic Latex. To render the cell, press shift+enter or run.

Alternatively, you can submit a pdf of your hand-written and scanned in answers to the theoretical questions on Learn, alongside this notebook.

# Question 1: Polynomial Regression

Consider the least squares regression problem introduced in lectures, where we wish to reconstruct an unknown function from noisy point values. Suppose we are given a set of points $\{x_i, y_i\}_{i=1}^m$, and we want to find the degree $n-1$ polynomial that best fits these points.  In other words, we want to learn the unknown coefficients $\{u_j\}_{j=1}^n$ in the polynomial
$$
f(x; u) = \sum_{j=1}^{n} {u_j} x^{j-1}, 
$$
such that the residual $\sum_{i=1}^m (y_i - f(x_i;u))^2$ is minimised.


### 1.1

Show that the regression problem above can be formulated as the linear inverse problem

$$
u^* = \text{argmin} \frac{1}{2} \|A u - y\|_2^2,
$$

where $y = \{y_i\}_{i=1}^m \in \mathbb R^m$ and $A \in \mathbb R^{m \times n}$ is the Vandermonde matrix with entries $a_{ij} = x_i^{j-1}$.

Since $(\mathbf{A}u-y)_i=\sum_{j=1}^n a_{ij}u_j-y_i$,
\begin{equation}
    \begin{aligned}
    \left\Vert\mathbf{A}u-y\right\Vert^2_2
    &=\sum_{i=1}^m \Big( \sum_{j=1}^n a_{ij}u_j-y_i\Big)^2\\
    &=\sum_{i=1}^m \Big( \sum_{j=1}^n x_i^{j-1}u_j-y_i\Big)^2\\
    &=\sum_{i=1}^m \Big( f(x_i;u)-y_i\Big)^2\\
    &=\sum_{i=1}^m \Big(y_i-f(x_i;u)\Big)^2.
    \end{aligned}
\end{equation}
Hence, 
\begin{equation}
    \begin{aligned}
    u^{*}&=\arg\min\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2\\
    &=\arg\min\left\Vert\mathbf{A}u-y\right\Vert^2_2\\
    &=\arg\min \sum_{i=1}^m \Big(y_i-f(x_i;u)\Big)^2.
    \end{aligned}
\end{equation}

### 1.2

Show that the gradient of $\frac{1}{2} \|A u - y\|_2^2$ with respect to $u$ is 

$$A^\mathrm T A u - A^\mathrm T y.$$

$\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2$ is the objective function.

Since $\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2=\frac{1}{2}\sum_{i=1}^m \Big( \sum_{j=1}^n a_{ij}u_j-y_i\Big)^2$, for $k=1,2,\ldots,n$, we have
\begin{equation}
    \begin{aligned}
    \Big(\frac{\partial}{\partial u}\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2\Big)_k
    &=\frac{\partial}{\partial u_k}\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2\\
    &=\frac{1}{2}\frac{\partial}{\partial u_k}\sum_{i=1}^m \Big( \sum_{j=1}^n a_{ij}u_j-y_i\Big)^2\\
    &=\frac{1}{2}\sum_{i=1}^m \frac{\partial}{\partial u_k}\Big( \sum_{j=1}^n a_{ij}u_j-y_i\Big)^2\\
    &=\frac{1}{2}\sum_{i=1}^m 2\Big(\sum_{j=1}^n a_{ij}u_j-y_i\Big)a_{ik}\\
    &=\sum_{i=1}^m a_{ik}\Big( \sum_{j=1}^n a_{ij}u_j-y_i\Big)\\
    &=\sum_{i=1}^m a_{ik} (\mathbf{A}u-y)_i
    \end{aligned}
\end{equation}
where $(a_{1k},\ldots,a_{mk})$ is the $k^{th}$ column vector of $\mathbf{A}$ and the $k^{th}$ row vector of $\mathbf{A^T}$, $(\mathbf{A}u-y)$ is a $m\times 1$ column vector. 

Let $(\mathbf{A^T})^{row}_k$ denote the $k^{th}$ row vector of $\mathbf{A^T}$, then
\begin{equation}
    \begin{aligned}
    \Big(\frac{\partial}{\partial u}\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2\Big)_k
    &= (\mathbf{A^T})^{row}_k (\mathbf{A}u-y).
    \end{aligned}
\end{equation}
Hence, 
\begin{equation}
    \frac{\partial}{\partial u}\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2=\mathbf{A^T}(\mathbf{A}u-y)= \mathbf{A^T}\mathbf{A}u-\mathbf{A^T}y.
\end{equation}

### 1.3 

In the code cell below, implement the method of gradient descent to compute a minimiser $u^*$. You should choose a suitable step size $h$.

Test your method on the simple, noise-free example given at the bottom of the code cell, to which the answer is $u = \begin{bmatrix} 1 \\ 2 \end{bmatrix}$.

In [1]:
import numpy as np

# add code here
def GDminimiser(y,A,h,u_0,num):
    #h is step size, u_0 is initial point, num is the number of iterations 
    """
    Return the minimiser obtained from gradiant descent method
    """
    u=u_0  #starting point
    k=1    #index for iteration
    
    for k in range(1,num): #run num interations
        #in each iteration
        #claculate the gradient of the objective function with respect to u
        delta_u=A.T.dot(A).dot(u)-A.T.dot(y) 
        #update the value of u via the gradiant method
        u=u-delta_u*h 
        
    return u #output is the minimiser after n iterations

# Testing the function on a small example
x = np.array([-0.5, 0, 0.5])
y = np.array([0., 1., 2.])
A = np.array([[1., -0.5],[1., 0.],[1., 0.5]])

u_0=np.array([0,0]) #set the initial value of u as (0,0)
#use a constant step size of 1/3
print("The minimiser from the method of gradient descent is",GDminimiser(y,A,1/3,u_0,1000) )


The minimiser from the method of gradient descent is [1. 2.]


### 1.4

In the code cell below, implement the Nestorov method both for convex and strongly convex objectives and  compute a minimiser $u^*$. 

Test your method on the simple, noise-free example given at the bottom of the code cell, to which the answer is $u = \begin{bmatrix} 1 \\ 2 \end{bmatrix}$.

In [2]:
import numpy as np

# add code here
#This is an optimisaion problem for least-squares estimator, so L equals to the largest eigenvalue 
#of A.T.dot(A), mu equals to the smallest eigenvalue of A.T.dot(A).
L=3     #L is Lipschitz constant of the first derivative of the objective function
mu=0.5  #the objective function is mu-strongly convex

def convex_min(y,A,u_0,num):
    #u_0 is the initial point
    """
    Return the minimiser obtained from Nestorov method for convex objectives
    """
    u=u_0  #set the initial value of u as u_0 
    v=u    #v will be updated after each iteration
    k=1    #index for iteration
    t_minus1=1  #initial value of a sequence values of t
    t=t_minus1  #current value of t
    
    for k in range(1,num): #run n interations
        #in each iteration
        t_plus=((1+4*(t**2))**(0.5)+1)/2   #calculate the new value for t
        gamma=(t-1)/t_plus                 #calculate the value of gamma
        t=t_plus                           #update the value of t
        #calculate the gradient of the objective function at the point v
        delta_v=A.T.dot(A).dot(v)-A.T.dot(y) 
        u_plus=v-delta_v/L          #calculate the new value for u
        v=u_plus+gamma*(u_plus-u)   #update the value of v
        u=u_plus                    #update the value of u
    
    return u #output is the minimiser after n iterations

def Sconvex_min(y,A,u_0,L,mu,num):
    #u_0 is the initial point
    """
    Return the minimiser obtained from Nestorov method for strongly convex objectives
    """
    u=u_0  #set the initial value of u as u_0 
    v=u    #v will be updated after each iteration
    gamma=(L**(0.5)-mu**(0.5))/(L**(0.5)+mu**(0.5)) #calculate the constant value of gamma
    k=1    #index for iteration
    
    for k in range(1,num): #run n interations
        #in each iteration
        #calculate the gradient of the objective function at the point v
        delta_v=A.T.dot(A).dot(v)-A.T.dot(y) 
        u_plus=v-delta_v/L          #calculate the new value for u
        v=u_plus+gamma*(u_plus-u)   #update the value of v
        u=u_plus                    #update the value of u
    
    return u #output is the minimiser after n iterations
        
# Testing the function on a small example
x = np.array([-0.5, 0, 0.5])
y = np.array([0., 1., 2.])
A = np.array([[1., -0.5],[1., 0.],[1., 0.5]])

u_0=np.array([0,0]) #set the initial value of u as (0,0)
print("The minimiser from Nestorov method for convex objectives is",convex_min(y,A,u_0,1000))
print("The minimiser from Nestorov method for strongly convex objectives is",Sconvex_min(y,A,u_0,3,0.5,1000))



The minimiser from Nestorov method for convex objectives is [1. 2.]
The minimiser from Nestorov method for strongly convex objectives is [1. 2.]


### 1.5

When $m \geq n$ (i.e. we are dealing with an overdetermined least squares problem) and $rank(A) = n$ (i.e. the matrix $A$ has full column rank), the minimisation problem in Question 1.1 is strongly convex, and hence has a unique minimiser $u^*$.

Compare the performance of gradient descent and Nestorov in computing $u^*$, for the data $\{x_i, y_i\}_{i=1}^m$ loaded in the code cell below. You may want to compare the computed solutions to the solution obtained with *np.linalg.lstsq(A,y)[0]*.

Include any numerical tests that you run in the code cell below. Any output you use in your discussion should be displayed and easily interpretable.

In [4]:
import numpy as np

# Load data points {x_i} - uniformly spaced points in [-1,1]
x = np.load('data_x.npy')

# Load corresponding data points {y_i}
y = np.load('data_y.npy')


# add your code here
m=x.size      #m is the number of data points
n=10          #n is the order of the polynomial degree
A=np.zeros(m*n)     #create a vector with m*n entries 
for i in range(1,m):     #for entries of the i^th row of matrix A
    for j in range(1,n): #for entries of the i^th column of matrix A
        #calculate the j^th component of the i^th row of matrix A
        A[(i-1)*n+j-1]=x[i-1]**(j-1) 
A=A.reshape((m,n)) #reshape A to be a m*n matrix
eigen=np.linalg.eigvals(A.T.dot(A)) #eigen is the vector of eigen values of A.T.dot(A)
L=eigen.max()  #Set the value of L the largest eigen value of A.T.dot(A)
mu=eigen.min() #Set the value of mu the smallest eigen value of A.T.dot(A)

##Set the number of iterations as 1000. 
u_0=np.zeros(n) #set the initial value of all elements of u as 0
#use a constant step size of 1/L
u_1=GDminimiser(y,A,1/L,u_0,1000)
print("The minimiser from the method of gradient descent after 1000 iterations is",u_1)

u_2=Sconvex_min(y,A,u_0,L,mu,1000)
print("The minimiser from Nestorov method for strongly convex objectives after 1000 iterations is",u_2)

u_3=np.linalg.lstsq(A,y,rcond=None)[0]
print("Solution obtained with np.linalg.lstsq(A,y)[0]",u_3)

def resiValue(u_estimate,x,y,m,n):
    """
    Return the residuals of different methods
    """
    res=0
    for i in range(1,m): #we will sum the residuals of all m data points
        ux=0  #ux is the fitted value of y
        for j in range(1,n):
            ux=ux+u_estimate[j-1]*(x[i-1]**(j-1))  #calculate the fitted value for each data point
        res=res+(y[i-1]-ux)**2  #cumulative residuals for the first i data points
    return res #output is the residuals

print("The residual of the method of gradient descent after 1000 iterations is",resiValue(u_1,x,y,m,n))
print("The residual of Nestorov method for strongly convex objectives after 1000 iterations is",resiValue(u_2,x,y,m,n))
print("The residual of using solution obtained with np.linalg.lstsq(A,y)[0]",resiValue(u_3,x,y,m,n))

##Set the number of iterations as 10000. 
u_0=np.zeros(n) #set the initial value of all elements of u as 0
#use a constant step size of 1/L
u_1=GDminimiser(y,A,1/L,u_0,10000)
print("The minimiser from the method of gradient descent after 10000 iterations is",u_1)

u_2=Sconvex_min(y,A,u_0,L,mu,10000)
print("The minimiser from Nestorov method for strongly convex objectives after 10000 iterations is",u_2)

u_3=np.linalg.lstsq(A,y,rcond=None)[0]
print("Solution obtained with np.linalg.lstsq(A,y)[0]",u_3)

print("The residual of the method of gradient descent after 10000 iterations is",resiValue(u_1,x,y,m,n))
print("The residual of Nestorov method for strongly convex objectives after 10000 iterations is",resiValue(u_2,x,y,m,n))
print("The residual of using solution obtained with np.linalg.lstsq(A,y)[0]",resiValue(u_3,x,y,m,n))
            
######Set the number of iterations as 100000. 
u_0=np.zeros(n) #set the initial value of all elements of u as 0
#use a constant step size of 1/L
u_1=GDminimiser(y,A,1/L,u_0,100000)
print("The minimiser from the method of gradient descent after 100000 iterations is",u_1)

u_2=Sconvex_min(y,A,u_0,L,mu,100000)
print("The minimiser from Nestorov method for strongly convex objectives after 100000 iterations is",u_2)

u_3=np.linalg.lstsq(A,y,rcond=None)[0]
print("Solution obtained with np.linalg.lstsq(A,y)[0]",u_3)

print("The residual of the method of gradient descent after 100000 iterations is",resiValue(u_1,x,y,m,n))
print("The residual of Nestorov method for strongly convex objectives after 100000 iterations is",resiValue(u_2,x,y,m,n))
print("The residual of using solution obtained with np.linalg.lstsq(A,y)[0]",resiValue(u_3,x,y,m,n))



The minimiser from the method of gradient descent after 1000 iterations is [-0.06549014 -0.19228465  0.26881368  0.89762087  0.38048177  0.60825961
 -0.15680148 -1.95149743 -0.82514545  0.        ]
The minimiser from Nestorov method for strongly convex objectives after 1000 iterations is [ -0.08024633   3.93926188   2.55264311 -34.69686622 -17.90092821
  77.11144624  38.71269974 -48.71937798 -24.76092292   0.        ]
Solution obtained with np.linalg.lstsq(A,y)[0] [ -0.10100278   2.39670315   2.57656176 -22.33974528 -15.80921523
  51.75181661  32.44369722 -33.71134564 -20.3108598    0.        ]
The residual of the method of gradient descent after 1000 iterations is 58.396071289072
The residual of Nestorov method for strongly convex objectives after 1000 iterations is 51.46112375714068
The residual of using solution obtained with np.linalg.lstsq(A,y)[0] 49.17687495533957
The minimiser from the method of gradient descent after 10000 iterations is [-0.02108256  0.02347249 -0.44548385 -1.8

It gives the smallest residual when we use solution obtained with np.linalg.lstsq(A,y)[0] as the estimated coefficients. When the number of iterations is small (1,000) or moderate (10,000), the residual of gradient descent method is larger than the residual of Nestorov method. When the number of iteration gets larger (100,000),  the residual of gradient descent method gets smaller than the residual of Nestorov method. Hence, in terms of residuals, Nestorov method method performs better in the beginning when the number of iteration is relatively small. As the number of iteration increases, gradient descent method performs better. 

# Question 2: Regularised Polynomial Regression 

When $m \leq n$ (i.e. we are dealing with an underdetermined least squares problem), the minimisation problem in Question 1.1 is convex.

Consider now the regularised minimisation problem

$$
u^* = \text{argmin} \frac{1}{2} \|A u - y\|_2^2 + \frac{\lambda}{2} \| u\|_2^2,
$$

which is strongly convex, and hence has a unique solution $u^*$, for any choice of regularisation parameter $\lambda > 0$.

### 2.1

Show that the gradient of $\frac{1}{2} \|A u - y\|_2^2 + \frac{\lambda}{2} \| u\|_2^2$ with respect to $u$ is 

$$(A^\mathrm T A + \lambda I) u - A^\mathrm T y.$$

Since derivatives are linear,
\begin{equation}
    \frac{\partial}{\partial u}\Big(\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2+\frac{\lambda}{2}\left\Vert u\right\Vert^2_2\Big)
    =\frac{\partial}{\partial u}\Big(\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2\Big)+\lambda\frac{\partial}{\partial u}\Big(\frac{1}{2}\left\Vert u\right\Vert^2_2\Big).
\end{equation}
From Question 1.2, we have 
\begin{equation}
    \frac{\partial}{\partial u}\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2=\mathbf{A^T}\mathbf{A}u-\mathbf{A^T}y,
\end{equation}
and $\frac{1}{2}\left\Vert u\right\Vert^2_2$ is a special case when $m=n$, $\mathbf{A}=\mathbb{1}$ and $y=0$, so
\begin{equation}
    \begin{aligned}
    \frac{\partial}{\partial u}\Big(\frac{1}{2}\left\Vert u\right\Vert^2_2\Big)
    &=\Bigg[\frac{\partial}{\partial u}\Big(\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2\Big)\Bigg]_{\mathbf{A}=\mathbb{1},y=0}\\
    &=\mathbb{1}^T\mathbb{1}u\\
    &=\mathbb{1}u.
    \end{aligned}
\end{equation}
Therefore, we have 
\begin{equation}
    \begin{aligned}
    \frac{\partial}{\partial u}\Big(\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2+\frac{\lambda}{2}\left\Vert u\right\Vert^2_2\Big)
    &=\frac{\partial}{\partial u}\Big(\frac{1}{2}\left\Vert\mathbf{A}u-y\right\Vert^2_2\Big)+\lambda\frac{\partial}{\partial u}\Big(\frac{1}{2}\left\Vert u\right\Vert^2_2\Big)\\
    &=\mathbf{A^T}\mathbf{A}u-\mathbf{A^T}y+\lambda\mathbb{1}u\\
    &=(\mathbf{A^T}\mathbf{A}+\lambda\mathbb{1})u-\mathbf{A^T}y.
    \end{aligned}
\end{equation}

### 2.2

Compare the performance of gradient descent and Nestorov in computing $u^*$ in the case $\lambda = 0$. You may want to compare your computed solution to a suitably obtained reference solution.

Include any numerical tests that you run in the code cell below. Any output you use in your discussion should be displayed and easily interpretable.

In [9]:
import numpy as np

# Load matrix A
A = np.load('data_A.npy')

# Load corresponding data points {y_i}
y = np.load('data_b.npy')


# add your code here
import scipy.linalg
from scipy.linalg import qr
import matplotlib.pyplot as plt
from scipy.optimize import nnls
m=A.shape[0] #m is the number of data points
n=A.shape[1] #n-1 is the polynomial degree
eigen=np.linalg.eigvals(A.T.dot(A)) #eigen is the vector of eigen values of A.T.dot(A)
L=eigen.max()  #Set the value of L the largest eigen value of A.T.dot(A)
mu=eigen.min() #Set the value of mu the smallest eigen value of A.T.dot(A)
print("The value of L is",L)
print("The value of mu is",mu)

def GDminimiser(y,A,h,u_0,num):
    #h is step size, u_0 is initial point, num is the number of iterations 
    """
    Return the minimiser obtained from gradiant descent method
    """
    u=u_0  #starting point
    k=1    #index for iteration
    
    for k in range(1,num): #run num interations
        #in each iteration
        #claculate the gradient of the objective function with respect to u
        delta_u=A.T.dot(A).dot(u)-A.T.dot(y) 
        #update the value of u via the gradiant method
        u=u-delta_u*h 
        
    return u #output is the minimiser after n iterations

def convex_min(y,A,u_0,num):
    #u_0 is the initial point
    """
    Return the minimiser obtained from Nestorov method for convex objectives
    """
    u=u_0  #set the initial value of u as u_0 
    v=u    #v will be updated after each iteration
    k=1    #index for iteration
    t_minus1=1  #initial value of a sequence values of t
    t=t_minus1  #current value of t
    
    for k in range(1,num): #run n interations
        #in each iteration
        t_plus=((1+4*(t**2))**(0.5)+1)/2   #calculate the new value for t
        gamma=(t-1)/t_plus                 #calculate the value of gamma
        t=t_plus                           #update the value of t
        #calculate the gradient of the objective function at the point v
        delta_v=A.T.dot(A).dot(v)-A.T.dot(y) 
        u_plus=v-delta_v/L          #calculate the new value for u
        v=u_plus+gamma*(u_plus-u)   #update the value of v
        u=u_plus                    #update the value of u
    
    return u #output is the minimiser after n iterations

def Sconvex_min(y,A,u_0,L,mu,num):
    #u_0 is the initial point
    """
    Return the minimiser obtained from Nestorov method for strongly convex objectives
    """
    u=u_0  #set the initial value of u as u_0 
    v=u    #v will be updated after each iteration
    gamma=(L**(0.5)-mu**(0.5))/(L**(0.5)+mu**(0.5)) #calculate the constant value of gamma
    k=1    #index for iteration
    
    for k in range(1,num): #run n interations
        #in each iteration
        #calculate the gradient of the objective function at the point v
        delta_v=A.T.dot(A).dot(v)-A.T.dot(y) 
        u_plus=v-delta_v/L          #calculate the new value for u
        v=u_plus+gamma*(u_plus-u)   #update the value of v
        u=u_plus                    #update the value of u
    
    return u #output is the minimiser after n iterations

##Set the number of iterations as 1000. 
u_0=np.zeros(n) #set the initial value of all elements of u as 0

u_1=GDminimiser(y,A,1/L,u_0,1000) #use a constant step size of 1/L
print("The minimiser from the method of gradient descent after 1000 iterations is",u_1)

u_2=Sconvex_min(y,A,u_0,L,mu,1000)
print("The minimiser from Nestorov method for strongly convex objectives after 1000 iterations is",u_2)

u_3=np.linalg.lstsq(A[:,0:m],y,rcond=None)[0]
print("Solution obtained with np.linalg.lstsq(A[:,0:m],y)[0]",u_3)

u_4=nnls(A,y)[0] #solution obtained with nnls(A,y)[0]



def resi_Value(u_estimate,A,y,m,n):
    """
    Return the residuals of different methods
    """
    res=0
    for i in range(1,m): #we will sum the residuals of all m data points
        res=res+(y[i-1]-u_estimate.dot(A[i-1]))**2
    return res #output is the residuals

print("The residual of the method of gradient descent after 1000 iterations is",resi_Value(u_1,A,y,m,n))
print("The residual of Nestorov method for strongly convex objectives after 1000 iterations is",resi_Value(u_2,A,y,m,n))
print("The residual of using solution obtained with np.linalg.lstsq(A[:,0:m],y)[0]",resi_Value(u_3,A[:,0:m],y,m,n))
print("The residual of using solution obtained with nnls(A,y)[0]",resi_Value(u_4,A,y,m,n))


The value of L is (512.0610267127927+0j)
The value of mu is (-1.597962496517407e-14+0j)
The minimiser from the method of gradient descent after 1000 iterations is [0.81347531+0.j 0.95854259+0.j 0.21190902+0.j 0.6124084 +0.j
 0.40598907+0.j 0.59461731+0.j 0.54123763+0.j 0.61337079+0.j
 0.61226184+0.j 0.62318392+0.j 0.64539801+0.j 0.62064641+0.j
 0.65843046+0.j 0.60940509+0.j 0.66108183+0.j 0.59328261+0.j
 0.65842308+0.j 0.57507233+0.j 0.65308891+0.j 0.55657743+0.j
 0.64646585+0.j 0.53887144+0.j 0.63929665+0.j 0.5225387 +0.j
 0.63198508+0.j 0.50785149+0.j 0.62475156+0.j 0.49489029+0.j
 0.61771474+0.j 0.48362225+0.j 0.61093556+0.j 0.47395119+0.j
 0.6044419 +0.j 0.46574896+0.j 0.59824264+0.j 0.45887471+0.j
 0.59233591+0.j 0.45318645+0.j 0.58671399+0.j 0.44854776+0.j
 0.58136615+0.j 0.44483135+0.j 0.57628033+0.j 0.44192077+0.j
 0.57144408+0.j 0.43971086+0.j 0.56684512+0.j 0.43810761+0.j
 0.5624716 +0.j 0.43702752+0.j 0.55831218+0.j 0.43639685+0.j
 0.55435612+0.j 0.43615067+0.j 0.55059322+0.

Compared with 1.5, residuals in 2.2 are much smaller. To obtain a reference solution, eatract the first m rows of matrix A to be new matrix, and then use np.linalg.lstsq(A,y)[0]. We can see that gradient descent method give similar coefficient estimators for every polynomial term of x. The coefficients obtained from Nestorov method for strongly convex objectives are more variant for the lower order polynomial terms and more similar for the higher order polynomial terms. The coefficients obtained from np.linalg.lstsq(A,y)[0] after slicing matrix A have very large magnitude. Both the gradient descent method and Nestorov method give similar residuals, which are also greater than the the residuals of using solution obtained with np.linalg.lstsq(A[:,0:m],y)[0]. So in terms of residuals, both methods have similar performance. However, overfitting might be a potential problem.

### 2.3

Compare the performance of gradient descent and Nestorov (both versions) in computing $u^*$ in the case $\lambda = 0.01\|A^T A\|_{2}$. In this case the unique solution can be obtained by solving an appropriate augmented least squares problem. 

Include any numerical tests that you run in the code cell below. Any output you use in your discussion should be displayed and easily interpretable.

In [3]:
import numpy as np

# Load matrix A
A = np.load('data_A.npy')

# Load corresponding data points {y_i}
y = np.load('data_b.npy')


# add your code here
m=A.shape[0] #m is the number of data points
n=A.shape[1] #n-1 is the polynomial degree
eigen=np.linalg.eigvals(A.T.dot(A)) #eigen is the vector of eigen values of A.T.dot(A)
L=eigen.max()  #Set the value of L the largest eigen value of A.T.dot(A)
mu=eigen.min() #Set the value of mu the smallest eigen value of A.T.dot(A)
print("The value of L is",L)
print("The value of mu is",mu)
u_0=np.zeros(n) #set the initial value of all elements of u as 0
lambd=0.01*np.linalg.norm(A.T.dot(A),ord=2)
print("the value of lambda is",lambd)
def GDminimiser2(y,A,h,u_0,num):
    #h is step size, u_0 is initial point, num is the number of iterations 
    """
    Return the minimiser obtained from gradiant descent method
    """
    u=u_0  #starting point
    k=1    #index for iteration
    
    for k in range(1,num): #run num interations
        #in each iteration
        #claculate the gradient of the objective function with respect to u
        delta_u=(A.T.dot(A)+lambd*np.identity(n)).dot(u)-A.T.dot(y) 
        #update the value of u via the gradiant method
        u=u-delta_u*h 
        
    return u #output is the minimiser after n iterations


def Sconvex_min2(y,A,u_0,L,mu,num):
    #u_0 is the initial point
    """
    Return the minimiser obtained from Nestorov method for strongly convex objectives
    """
    u=u_0  #set the initial value of u as u_0 
    v=u    #v will be updated after each iteration
    gamma=(L**(0.5)-mu**(0.5))/(L**(0.5)+mu**(0.5)) #calculate the constant value of gamma
    k=1    #index for iteration
    
    for k in range(1,num): #run n interations
        #in each iteration
        #calculate the gradient of the objective function at the point v
        delta_v=(A.T.dot(A)+lambd*np.identity(n)).dot(u)-A.T.dot(y)
        u_plus=v-delta_v/L          #calculate the new value for u
        v=u_plus+gamma*(u_plus-u)   #update the value of v
        u=u_plus                    #update the value of u
    
    return u #output is the minimiser after n iterations

def convex_min2(y,A,u_0,num):
    #u_0 is the initial point
    """
    Return the minimiser obtained from Nestorov method for convex objectives
    """
    u=u_0  #set the initial value of u as u_0 
    v=u    #v will be updated after each iteration
    k=1    #index for iteration
    t_minus1=1  #initial value of a sequence values of t
    t=t_minus1  #current value of t
    
    for k in range(1,num): #run n interations
        #in each iteration
        t_plus=((1+4*(t**2))**(0.5)+1)/2   #calculate the new value for t
        gamma=(t-1)/t_plus                 #calculate the value of gamma
        t=t_plus                           #update the value of t
        #calculate the gradient of the objective function at the point v
        delta_v=(A.T.dot(A)+lambd*np.identity(n)).dot(u)-A.T.dot(y)
        u_plus=v-delta_v/L          #calculate the new value for u
        v=u_plus+gamma*(u_plus-u)   #update the value of v
        u=u_plus                    #update the value of u
    
    return u #output is the minimiser after n iterations

def resi_Value(u_estimate,A,y,m,n):
    """
    Return the residuals of different methods
    """
    res=0
    for i in range(1,m): #we will sum the residuals of all m data points
        res=res+(y[i-1]-u_estimate.dot(A[i-1]))**2
    return res #output is the residuals

##set the number of iterations as 1000
uu_1=GDminimiser2(y,A,1/L,u_0,1000) #use a constant step size of 1/L
#print("The minimiser from the method of gradient descent after 1000 iterations is",uu_1)
print("The residual of the method of gradient descent after 1000 iterations is",resi_Value(uu_1,A,y,m,n))

uu_2=Sconvex_min2(y,A,u_0,L,mu,1000)
#print("The minimiser from Nestorov method for strongly convex objectives after 1000 iterations is",uu_2)
print("The residual of Nestorov method for strongly convex objectives after 1000 iterations is",resi_Value(uu_2,A,y,m,n))

uu_3=convex_min2(y,A,u_0,1000)
#print("The minimiser from Nestorov method for strongly convex objectives after 1000 iterations is",uu_3)
print("The residual of Nestorov method for convex objectives after 1000 iterations is",resi_Value(uu_3,A,y,m,n))

##set the number of iterations as 10000
uu_1=GDminimiser2(y,A,1/L,u_0,10000) #use a constant step size of 1/L
#print("The minimiser from the method of gradient descent after 10000 iterations is",uu_1)
print("The residual of the method of gradient descent after 10000 iterations is",resi_Value(uu_1,A,y,m,n))

uu_2=Sconvex_min2(y,A,u_0,L,mu,10000)
#print("The minimiser from Nestorov method for strongly convex objectives after 10000 iterations is",uu_2)
print("The residual of Nestorov method for strongly convex objectives after 10000 iterations is",resi_Value(uu_2,A,y,m,n))

uu_3=convex_min2(y,A,u_0,10000)
#print("The minimiser from Nestorov method for strongly convex objectives after 10000 iterations is",uu_3)
print("The residual of Nestorov method for convex objectives after 10000 iterations is",resi_Value(uu_3,A,y,m,n))

##set the number of iterations as 100000
uu_2=Sconvex_min2(y,A,u_0,L,mu,100000)
#print("The minimiser from Nestorov method for strongly convex objectives after 100000 iterations is",uu_2)
print("The residual of Nestorov method for strongly convex objectives after 100000 iterations is",resi_Value(uu_2,A,y,m,n))


The value of L is (512.0610267127927+0j)
The value of mu is (-1.597962496517407e-14+0j)
the value of lambda is 5.120610267127922
The residual of the method of gradient descent after 1000 iterations is (1.8288473560713514+0j)
The residual of Nestorov method for strongly convex objectives after 1000 iterations is (49059.809178733376-0.3776792958712106j)
The residual of Nestorov method for convex objectives after 1000 iterations is (1.8300432291179611+0j)
The residual of the method of gradient descent after 10000 iterations is (1.828846908029326+0j)
The residual of Nestorov method for strongly convex objectives after 10000 iterations is (4463.086357838996-1.6318141712968444j)
The residual of Nestorov method for convex objectives after 10000 iterations is (1.8288129469819212+0j)
The residual of Nestorov method for strongly convex objectives after 100000 iterations is (4836.5820351627635-7.477858911311769j)


When the number of iterations is 1,000 or 10,000, the residual(around 1.82) of gradient descents method is close to the residual(around 1.83) of Nestorov method for convex objectives. Both residuals are much smaller than the residual of Nestorov method for strongly convex objectives. In terms of residuals, both gradient descents method and Nestorov method for convex objectives perform well. In terms of the time to implement the method with the same number of iterations, the gradient method performs better. Also, when the number of iteration increases (to 10,000), the residual of Nestorov method for strongly convex objectives decreases from 49059 to 4463. However, the residual is still very large (4836 after 100,000 iterations) and it takes long time to implement Nestorov method for strongly convex objectives when the number of iterations is large.